### ISSUE TITLE
add jwt to header

### ISSUE THREAD

**Issue by richard-ellison** [ElSnoMan/pyleniumio#232]
Hello,

First, off thanks! The small amount of code to get a basic example going is refreshing compared with all the faff I've had in the past with using selenium directly.

I can't see an example of how to do this in the videos or the docs. If someone can point me in the right direction I'll get a working example going and add it in via a PR if that's helpful?

Looking to add E2E automated testing for a web application that uses firebase JWT in the header for auth. In my unit tests for the app, I have got a session scoped pytest fixture that logs in once via a request API call. I then pass this header around to the tests that need it. 

I am currently looking for how I can do the same for a separate set of integration tests using pyleniumio. My thought process is that I could add the jwt into the py fixture so that it uses it in every header. Doing so via a session scoped pytest fixture so I only have to login once during a test run. Otherwise, firebase will naturally get upset that I am logging in separately for every test. 

Any pointers would be appreciated.

**Comment by ElSnoMan** [comment]
Hi, @richard-ellison!

Thanks for using Pylenium and creating an Issue!

When you get the JWT token for the website, is it stored as a cookie? Or how does the website persist the auth for a session?

**Comment by richard-ellison** [comment]
Hi @ElSnoMan 

Thanks for your reply.

Unfortunately it's not a cookie but an added header called authorisation with the value of the JWT.

My current reading via selenium docs and stackoverflow is pointing me in the direction of browsermob. If I understand correctly the work around is to manipulate the requests headers coming from the web driver via a proxy.

The selenium core Devs don't appear to support this and are saying that they only want to mimic users. Going on to say that tests for headers like this should be done by other software. 

So selenium and therefore pylenium wouldn't support adding in a header directly ?

github.com/lightbody/browsermob-proxy

I wonder if my testing approach is wroung, all I really want to do is login once and use that session for many seperate tests. Otherwise I think I am looking at one long test instead of many seperate ones.

If the "py" fixture was session scoped I could login there and pass that to each of my seperate tests. (Not saying it should be, just saying it would solve this case). What's the rational for py fixture being function scoped? is it worth me looking at a PR for making it and it's dependents session based or is that a design choice ?

Would very much appreciate your thoughts.

**Comment by ElSnoMan** [comment]
## Answers

> **So Selenium and therefore Pylenium wouldn't support adding in a header directly?**

I've never used something like `browsermob-proxy` before cuz I haven't had to, but if we could design a generalized solution for it, I wouldn't mind adding it as a feature to Pylenium.

> **What's the rationale for py fixture being function scoped?**

It's best practice to have your tests modular and atomic. In Web Automation, you don't want to open one browser and run 10 test functions against it. Instead, each test focuses on specific behavior and gets its own browser. This is for a lot of reasons including:
- run tests in parallel to reduce run times
- tests are not dependent on each other (no shared test data, etc)
- run tests separately for smarter executions
- etc

> 💡 "best practice" is more of a guideline of course. Different contexts and scenarios may call for different approaches.

For example, let's say we have 5 tests [A, B, C, D, E]. If they were modular and atomic, I could run all 5 concurrently or only a subset of them.

```
A   B   C   D   E
^   ^   ^   ^   ^
```

If the tests share a single browser, then I must run them in the correct order meaning that I can no longer run them concurrently.

```
A -> B -> C -> D -> E
```

If test C fails, then tests D and E are not executed. If I want to rerun test C, then I have to run A and B first.

## A Solution

With that said, the `Pylenium` class is just an object, so you can instantiate it and share between tests. You should make a custom fixture at the session scope. Something like this:

```python
@pytest.fixture(scope='session')
def pys(py_config, request):
    py = Pylenium(py_config)
    yield py
    py.quit()
```

Then use the fixture in your tests:

```python
def test_a(pys: Pylenium):
    ...

def test_b(pys: Pylenium):
    ...
```

**Comment by richard-ellison** [comment]
Hi @ElSnoMan ,

Thanks for the detailed response,

I have worked with pytest-xdist (was very happy to see you supported it) before and created separate databases for each cpu. Now that I look back at my code I can see now that the fixture going into the tests is session scope to ensure the db session is clean for the next tests use. 

So for the ideal test separation there would be a freshly logged in py as a function scoped fixture. 

The session scoped method looks like the way forward at lest for now, however I understand that it removes an element of separation. 

I did find a library called Selenium Wire which supports header manipulation so I might explore what they do to selenium to make it accept headers and see if a PR is appropriate for pylenium.io. This would then allow for a fresh py fixture to be used each time avoiding any cross contamination.

In the mean time once I have the session based option working do you want a PR for a Firebase auth example?

Kind Regards
Rich

**Comment by ElSnoMan** [comment]
I would love that! We welcome all ideas and PRs and I think a simple solution would go a long way for anyone else who needs this functionality.

Looking over the browsermob proxy repo, it seems like we may have to do some `pylenium.json` and `webdriver_factory` changes, but I'm interested to see how you solve it.

Thanks, @richard-ellison! I can't wait to see your auth example 😄

**Comment by ElSnoMan** [comment]
@richard-ellison SeleniumWire is integrated natively within Pylenium and we have another issue (#248) to create a session-based Pylenium instance and a class-based Pylenium instance.

I'm thinking of closing this Issue unless there's something specific you'd like to keep this open for. Let me know!

Thanks

**Comment by richard-ellison** [comment]
Hi @ElSnoMan, thank you for taking the time to message me. I look forward to the new functionality.
